operatorcourier/flatten.py

```diff
@@ -0,0 +1,151 @@
+import logging
+import os
+from yaml import safe_load, MarkedYAMLError
+import semver
+import ntpath
+import operatorcourier.identify as identify
+import operatorcourier.errors as errors
+
+logger = logging.getLogger(__name__)
+
+
+def get_flattened_files_info(source_dir: str) -> [(str, str)]:
+    """
+    :param source_dir: Path of the directory containing different versions
+    of operator bundles (CRD, CSV, package) in separate version directories
+    :return: A list of tuples where in each tuple, the first element is
+    the source file path to be copied to the flattened directory, and the second is
+    the new file name to be used in the file copy. The function returns an empty list if
+    the source_dir is already flat
+    """
+
+    # extract package file and version folders from source_dir
+    _, folder_names, file_names = next(os.walk(source_dir))
+    if len(folder_names) == 0:
+        logger.info('The source directory is already flat. Exiting the program...')
+        return []
+
+    file_paths_to_copy = []  # [ (SRC_FILE_PATH, NEW_FILE_NAME) ]
+
+    crd_dict = {}  # { CRD_NAME => (VERSION, CRD_PATH) }
+    csv_paths = []
+
+    for version_folder_name in folder_names:
+        parse_version_folder(source_dir, version_folder_name, csv_paths, crd_dict)
+
+    # 1. add package in source_dir
+    package_path, package_name = get_package_file(source_dir, file_names)
+    file_paths_to_copy.append((package_path, package_name))
+
+    # 2. add all CRDs with the latest version
+    for crd_name, (_, crd_path) in crd_dict.items():
+        file_paths_to_copy.append((crd_path, ntpath.basename(crd_path)))
+
+    # 3. add all CSVs
+    for csv_file_name, csv_entries in create_csv_dict(csv_paths).items():
+        if len(csv_entries) == 1:
+            file_paths_to_copy.append((csv_entries[0][1], csv_file_name))
+            continue
+        # duplicate CSVs found in different versions
+        for (version, csv_path) in csv_entries:
```

Comment:
Why not do this transformation also when `len(csv_entries) == 1`? And then the if-block above could be deleted.

Reply:
@csomh Without the `if` block, all CSVs with be appended with the version number. This part of the code is actually handling a corner case where two or more CSVs have the same file name. But in reality, we do not expect this to happen a lot (could you @kevinrizza confirm this?), and doing such transformation may usually end up duplicate version numbers in the file name.

Reply:
Are duplicate version numbers and issue?

Reply:
The filenames don't make a difference once we bundle, so I think it's fine if we don't have more complex logic here and instead just assume we want to rename the files.